### PR TITLE
feat(web, server): Ability to use config file instead of admin UI

### LINKED ISSUE #3774: Config file support
It would be nice to add support for an optional config file, so that every aspect of Immich can be configured without needing to use the GUI. This could be implemented as follows:

For simplicity, the config file should be in the JSON format and the schema should match [the existing internal configuration interface](https://github.com/immich-app/immich/blob/main/server/src/infra/entities/system-config.entity.ts#L93-L131), with all values being optional. That way, we don't need to do any additional format parsing or transformation.

The config file can be enabled by setting an environment variable, for example `CONFIG_FILE=/config/immich.json`. 

If the config file is set, The [`SystemConfigCore.getConfig()`](https://github.com/immich-app/immich/blob/main/server/src/domain/system-config/system-config.core.ts#L93C16-L102) function should use the contents of this config file instead of the configuration that is in the database. The admin UI should also be greyed out to not allow updating the database config, and display a message that Immich is configured via the config file.

Bonus: Since the config file format/schema is the same as that used by the API, we could add a button in the admin UI that lets you copy the currently active config, so you can easily move that into a config file.

### PR DESCRIPTION
**Pull request by danieldietzler**
This PR closes #3774, providing the ability to use a JSON config file instead of the admin UI.

Currently, the admin UI and the config file are mutually exclusive while the config file always takes precedence. So, if a config file is specified, changes in the admin UI are not possible anymore.
To get people started with this quickly, there is a new copy to clipboard button that allows to copy the entire config as a formatted JSON string.

If also added documentation for the new env variable, but I wasn't sure where to put the documentation for the config file.

Screenshot 2023-08-23 at 11 39 01

Screenshot 2023-08-23 at 11 34 31

### CODE REVIEW

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (mertalev) [inline server/src/domain/system-config/dto/system-config.dto.ts:6]:
```suggestion
import { toBoolean } from '../../domain.util';
```

**Review by mertalev — CHANGES_REQUESTED** [review]
Thanks for picking this up!

**Comment by jrasm91** [comment]
This looks really good. Looking at it now, I'm wondering if maybe it would be better to move the read config file method to the config repository instead of the storage repository, since that would reduce a lot of unnecessary changes in this PR.

**Comment by danieldietzler** [comment]
Yeah, that was quite an adventure... :D
I asked about this beforehand (in the issue) and this was suggested. This is my first suggestion, so I am really new to the code base and don't know where methods make the most sense.

So, if you'd say that it would make more sense in the config repository, I sure can change it! This would make this PR a lot slimmer and simpler ;)
Alternatively, we can also wait for others to chime in, maybe there are some people who have opinions on this.

**Comment by danieldietzler** [comment]
At least I got to know the code base a bit by this ^^

**Comment by jrasm91** [comment]
Yeah, I don't think there are strong preferences either way. I'm probably the primary developer in this area of the codebase and I'm good either way. I think it would be simpler myself and thought of it after posting my comment yesterday. Usually it is nice to keep related functions together, so reading a file naturally could belong with all the other file operations in the storage repository, but in this case that makes that change quite unwieldy and I don't see much benefit requiring it to be there. I usually prefer simplicity and moving it to the config repository would make it considerably simpler.

**Comment by danieldietzler** [comment]
On the other hand, now that I've refactored the method to read files in general, if there should be some other use case at some time where reading files would be needed again, the logic would then already be in the right place. So, it would make the life of future devs easier  :)

**Comment by jrasm91** [comment]
True although, it is trivial to add that again if needed 😛

**Comment by danieldietzler** [comment]
Sure, but this would result in duplicate code unless you merge the config file reading then, at which point you would have a bit more to do.

File: server/src/domain/system-config/system-config.core.ts
Comment (bo0tzz) [inline server/src/domain/system-config/system-config.core.ts:103]: We should probably validate the contents of the config file in some way and apply them on top of the defaults, rather than just blindly loading and accepting it. At the moment it would be possible to set unknown/unsupported values, and to neglect setting values that are required.

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (bo0tzz) [inline server/src/domain/system-config/dto/system-config.dto.ts:45]: At the moment I think it's technically possible to write to this value via the API (which would persist it to the database, which would probably mean a bad time). We should prevent that if we can.

**Review by bo0tzz — CHANGES_REQUESTED** [review]
Thanks for picking this up!

> I wasn't sure where to put the documentation for the config file.

This would probably suit a new page under `docs/docs/install`, after the env vars page.

**Comment by bo0tzz** [comment]
As for the file reading discussion: Reading a file from disk is a relatively trivial event, so I don't think we need to anticipate much future need there. It's probably best to just take the simplest approach/with the least changes needed.

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (danieldietzler) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: I know, yeah... But tbh I couldn't think of a way to avoid it. You could have different object models for requests and responses, but I don't think this is good practice... I guess the same issue turns up when manually deleting the isConfigFile value before pasting it to the clipboard in the front end.

File: server/src/domain/system-config/system-config.core.ts
Comment (danieldietzler) [reply server/src/domain/system-config/system-config.core.ts:103]: In system-config.core.ts the updateConfig function uses some validators that it gets injected. I guess I could use these, but I cannot find, how they are being injected and how I can get them? addValidator is only called by tests?

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (danieldietzler) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: So that the flag would the fetched separately?

Comment (mertalev) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: `updateConfig` uses `Object.values(SystemConfigKey)` to determine what gets stored in the database. Since `isConfigFile` isn't in `SystemConfigKey`, I don't think it would get persisted to the database either way.

Personally though, I would change the typing here so `SystemConfigDto` is unchanged and make a new type that contains `SystemConfigDto` and `isConfigFile` as fields. This clarifies what `isConfigFile` is and isn't, both to the compiler and to contributors.

File: server/src/domain/system-config/system-config.core.ts
Comment (mertalev) [reply server/src/domain/system-config/system-config.core.ts]: I haven't tested this, but I believe you can convert the object to a class and call `class-validator`'s `validate` in `readConfigFile`:

```suggestion
        const config = plainToClass(SystemConfigDto, JSON.parse(await fs.readFile(path, 'utf-8')))
        await validate(config, { forbidNonWhitelisted: true, forbidUnknownValues: true })
```

**Review by mertalev — DISMISSED** [review]
I think it's good once `isConfigFile` and validation are sorted out.

**Comment by alextran1502** [comment]
Hello, can you please include a screenshot of the UI elements?

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (danieldietzler) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: So there would be an extra SystemConfigResponse model? 
Having `isConfigFile` as a header flag is bad practice, right?

File: server/src/domain/system-config/system-config.core.ts
Comment (danieldietzler) [reply server/src/domain/system-config/system-config.core.ts]: Unfortunately, this does not validate invalid numbers for instance. Setting a wrong resolution for e.g. Thumbnails still yields an empty field in the UI but no error.

Comment (danieldietzler) [reply server/src/domain/system-config/system-config.core.ts]: I've just seen that valid values in selects are currently hardcoded in the UI. So I'd assume the specific values aren't being verified at all? Maybe the backend should verify them and provide endpoints to fetch valid options? 'Cause technically this isn't something the frontend should take care of, especially because it would yield duplicate code.

Comment (mertalev) [reply server/src/domain/system-config/system-config.core.ts]: The server rejects the updated config if any of the options fail the validation decorators in the DTOs. This is done in the system config controller where there's a `@UseValidation` decorator telling Nest.js to validate requests and responses. 

But some things essentially aren't validated since they're strings.

Comment (danieldietzler) [reply server/src/domain/system-config/system-config.core.ts]: Yeah, I got that. But I suppose values like resolution could use `@IsIn()` with a list of valid options. Wouldn't this be even tighter validation. This list could then be exposed to the frontend as well, so that the select options can be generated from it.

Comment (mertalev) [reply server/src/domain/system-config/system-config.core.ts]: Yeah, adding validation for that would be a good change. I've been slowly migrating those settings into explicit enums but never got around to changing those.

Generating the options from the type would be elegant, but it removes some flexibility. For instance, displayed text isn't always the same as the value, and there's no guarantee that all options will always be user-facing.

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (jrasm91) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: We can use a feature flag instead of this extra response setting imo.

File: server/src/domain/system-config/system-config.core.ts
Comment (danieldietzler) [reply server/src/domain/system-config/system-config.core.ts]: I feel like the config file and the UI should always have feature parity and, since the config file would be validated against all possible values, they should all be accessible from the UI, too.

Nevertheless, for now, this suffices as validation and we pick this up later on?

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (danieldietzler) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: So we wait for your PR to be merged (what's the current state?), then we use the feature here and merge this PR?

**Comment by danieldietzler** [comment]
I don't know why this test is failing because I haven't touched any of those files

**Comment by danieldietzler** [comment]
Why did the intl version change just from running `make api`? I don't think I should change the version in the `pubspec.yaml`, right?

File: server/src/domain/system-config/system-config.core.ts
Comment (mertalev) [reply server/src/domain/system-config/system-config.core.ts]: Agreed, I think this is fine.

**Comment by mertalev** [comment]
Not sure, but your openapi-generator version might be different than what's on main. I'd run `npm ci` in `server` and try again.

**Comment by jrasm91** [comment]
I think this is actually pinned in a patch file that probably isn't getting applied properly.

File: server/src/domain/system-config/dto/system-config.dto.ts
Comment (bo0tzz) [reply server/src/domain/system-config/dto/system-config.dto.ts:45]: It's just been merged, so that should be available now.

**Comment by danieldietzler** [comment]
> Not sure, but your openapi-generator version might be different than what's on main. I'd run `npm ci` in `server` and try again.

That unfortunately did not do the trick.

**Comment by danieldietzler** [comment]
> I think this is actually pinned in a patch file that probably isn't getting applied properly.

Sorry, but I can't seem to find any patch file that would be related to this. Also, they are generated and should not be touched either, right?
What to you propose I try, or are you debugging yourself already and trying to fix the issue?

File: server/src/domain/server-info/server-info.dto.ts
Comment (bo0tzz) [inline server/src/domain/server-info/server-info.dto.ts:83]: Nitpick: I think `usingConfigFile` or just `configFile` would be better naming for a feature flag.

**Review by bo0tzz — APPROVED** [review]

**Comment by jrasm91** [comment]
Added a few things:
- Renamed to `IMMICH_CONFIG_FILE`
- Show a notification after clicking copy to clipboard
- Add a button to export to json
- Validate the file (after merging it with the default configuration). This also throws errors for unknown keys
- Always show copy/export buttons, even when config is disabled.

**Review by jrasm91 — APPROVED** [review]

**Comment by danieldietzler** [comment]
Thanks a lot! :)

**Comment by bo0tzz** [comment]
Thank you for the PR!

**Pull request merged**
